jit-compiler/src/lib.rs

```diff
@@ -0,0 +1,54 @@
+mod codegen;
+mod compiler;
+
+use std::{collections::HashMap, fs};
+
+use codegen::CodeGenerator;
+use compiler::{call_cargo, generate_glue_code, load_library};
+use powdr_ast::analyzed::Analyzed;
+use powdr_number::FieldElement;
+
+pub type SymbolMap = HashMap<String, fn(u64) -> u64>;
+
+/// Compiles the given symbols (and their dependencies) and returns them as a map
+/// from symbol name to function pointer.
+/// Only functions of type (int -> int) are supported for now.
+pub fn compile<T: FieldElement>(
+    analyzed: &Analyzed<T>,
+    requested_symbols: &[&str],
+) -> Result<SymbolMap, String> {
+    log::info!("JIT-compiling {} symbols...", requested_symbols.len());
+
+    let mut codegen = CodeGenerator::new(analyzed);
+    let successful_symbols = requested_symbols
+        .iter()
+        .filter_map(|&sym| {
+            if let Err(e) = codegen.request_symbol(sym) {
+                log::warn!("Unable to generate code for symbol {sym}: {e}");
+                None
+            } else {
+                Some(sym)
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if successful_symbols.is_empty() {
+        return Ok(Default::default());
+    };
+
+    let glue_code = generate_glue_code(&successful_symbols, analyzed)?;
+
+    let (dir, lib_path) = call_cargo(&format!("{glue_code}\n{}\n", codegen.compiled_symbols()))?;
+    let metadata = fs::metadata(&lib_path).unwrap();
+
+    log::info!(
+        "Loading library of size {} MB...",
+        metadata.len() as f64 / 1000000.0
+    );
+
+    let result = load_library(&lib_path, &successful_symbols);
+    log::info!("Done.");
+
+    drop(dir);
```

Comment:
It's a temporary directory that is deleted when the reference is dropped. We need to keep it alive at least until the dlopen call.

Reply:
But the effect of drop here must be exactly the same as letting the function return. Did you mean `std::mem::forget()`, or even `dir.release()`?

If it is working with the drop, it should be working without it as well, and there is something else going on that allows dlopen to find the file.

Reply:
Ah, didn't see `dir.release()`. But at least `dir.release()` must be called, otherwise (as far as I understand), the compiler can drop the value any time even before the function returns.

Reply:
Yeah, I tried just removing this line and it still works.

Reply:
I don't understand how it works, either, but this drop is a nop 😎

Reply:
It actually did not work for me in the sense that `let (dir, ..) = ...` will lead to a clippy error because `dir` is unused. And if I remove `dir`, then the directory will be deleted.

Reply:
Ah, that explains the need for it: so that dir is used and clippy does not complain. Alternatively you can try naming it _dir.